### PR TITLE
Operator refactor  part 2: moving stuff out of the SWE-specific operators

### PR DESCRIPTION
**Pull request by jeff-cohere**
This PR implements some new data structures for accessing operator data in both CEED and PETSc. There are now specific types for accessing operator data for

* boundary values (fluxes)
* source terms
* material properties
* flux divergences communicated between the flux and source operators

These types pertain to operators for all kinds of conservative flow/transport equations that we might solve, so they can manipulate data for non-SWE operators. Accordingly, they replace their SWE-specific equivalents, allowing operator implementations to focus only on their specific physics. I intend these types also to handle things for both PETSc and CEED runs, whch localizes a lot of logic that differentiates between these two sets of solvers.

The data access structures work very similarly to how data is manipulated in PETSc/CEED vectors. You get access to operator data, set whatever values you want, and then release ("restore") the data in the operators. We allow only one entity at a time to access the data, which is fine, since "host calculations" within an MPI process are done completely serially.

In the next stage of the refactor, I'll attempt to replace SWE PETSc logic with these types, and patch in the flux divergences, which I haven't done yet.

### CODE REVIEW

**Comment by jeff-cohere** [comment]
Aha, code coverage! Yep, I added some code that hasn't been used yet. I don't think this is a real problem, since coverage will bounce back up after the next PR.

File: driver/main.F90 [hunk 1]
```diff
@@ -152,9 +152,6 @@ program rdycore_f90
       PetscCallA(RDyGetLocalCellZCentroids(rdy_, n, values, ierr))
       PetscCallA(RDyGetLocalCellNaturalIDs(rdy_, n, nat_id, ierr))
 
-      PetscCallA(RDyGetLocalCellManningsNs(rdy_, n, values, ierr))
-      PetscCallA(RDySetManningsNForLocalCells(rdy_, n, values, ierr))
```
Comment (bishtgautam) [inline driver/main.F90:156]: Will you add the capabilities to get/set Mannings N from a Fortran code in a future PR?

File: src/operator_data.c [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    PetscReal *u = boundary_data->storage.petsc.data;
+    for (PetscInt e = 0; e < boundary_data->boundary.num_edges; ++e) {
+      u[e * boundary_data->num_components + component] = boundary_values[e];
+    }
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+PetscErrorCode RestoreOperatorBoundaryData(RDy rdy, RDyBoundary boundary, OperatorBoundaryData *boundary_data) {
```
Comment (bishtgautam) [inline src/operator_data.c:133]: A few functions are missing the comment describing the function, but it can be added in a future PR.

**Review by bishtgautam — COMMENTED** [review]
This looks good and I have two minor comments.

File: driver/main.F90 (hunk 1 above)
Comment (jeff-cohere) [reply driver/main.F90:156]: This didn't actually do anything, as far as I can tell--it just copied the mannings values out and back in again (right?).

File: src/operator_data.c (hunk 2 above)
Comment (jeff-cohere) [reply src/operator_data.c:133]: Yeah, that sounds good.

File: driver/main.F90 (hunk 1 above)
Comment (bishtgautam) [reply driver/main.F90:156]: Yes, they only extracted or set values for ManningN. These functions could be needed for coupling with E3SM.

Comment (bishtgautam) [reply driver/main.F90:156]: These functions can be added back in a future PR.

**Review by bishtgautam — APPROVED** [review]

**Pull request merged**